libtenzir/builtins/operators/deduplicate.cpp

```diff
@@ -122,33 +136,143 @@ class deduplicate_operator final : public crtp_operator<deduplicate_operator> {
   using match_store
     = std::unordered_map<data, match_type, match_hash, std::equal_to<>>;
 
-  auto make_indices(const type& schema) const -> std::vector<offset> {
+  auto make_projection(const table_slice& slice, diagnostic_handler& diag) const
+    -> std::optional<cached_projection> {
     std::vector<offset> indices{};
+    std::vector<std::string> missing_fields{};
+    const auto& schema = slice.schema();
     for (const auto& field : cfg_.fields) {
+      bool resolved = false;
       for (auto idx : schema.resolve(field)) {
         indices.emplace_back(std::move(idx));
+        resolved = true;
+      }
+      if (not resolved) {
+        TENZIR_ASSERT(not field.empty());
+        if (field.starts_with(':')) {
+          // We can't easily deal with missing type extractors, just erroring out.
+          diagnostic::error(
+            "failed to deduplicate due to unmatched type extractor")
+            .note("`{}` matched no fields in schema `{}`", field, schema.name())
+            .emit(diag);
+          return std::nullopt;
+        }
+        missing_fields.emplace_back(field);
       }
     }
-    std::sort(indices.begin(), indices.end());
-    indices.erase(std::unique(indices.begin(), indices.end()), indices.end());
-    return indices;
+    if (missing_fields.empty()) {
+      // Every field in `cfg_.fields` was found in `slice.schema()`.
+      // Clean up the indices and return.
+      std::ranges::sort(indices);
+      {
+        auto uniq = std::ranges::unique(indices);
+        indices.erase(uniq.begin(), uniq.end());
+      }
+      return cached_projection{
+        .indices = std::move(indices),
+        .transformation_factory = {},
+      };
+    }
+    // Some fields were missing.
+    // Construct a `transformation_factory`, that'll add these fields in, with a
+    // value of `null`.
```

Comment:
IIUC, this also modifies the events returned by the operator, right? Not sure how I feel about this. Did you consider the alternative of not modifying the incoming events?

Reply:
The events returned by the operator are not modified -- the outputs are yielded in `deduplicate()`, and are only ever subslices of the original input. This null-trickery is part of the logic for comparing whether two events are duplicates of one another. Essentially, heavily simplified, the logic is:
```
// inside deduplicate::operator():
std::unordered_map<data, ...> matches{};
for (auto&& slice : input) {
  auto projected_slice = project(slice);

  size_t begin_row = 0;
  for (size_t current_row = 0; current_row < slice.rows(); ++current_row) {
    // .row_at() is pseudocode
    auto match_it = matches.find(projected_slice.row_at(current_row));
    if (match_it == matches.end()) {
      std::tie(match_it, std::ignore)
        = matches.emplace(projected_slice.row_at(current_row), ...);
    }
    if (is_duplicate(match_it)) {
      co_yield subslice(slice, begin_row, current_row);
      begin_row = current_row + 1;
    }
  }
  co_yield subslice(slice, begin_row, slice.rows());
}
```
The output of `project` is used for lookup into `matches`, i.e., to find out whether the row is a duplicate. The actual output contains only subslices of the original input.

Reply:
Thank you for the explanation :pray: